src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinderProvider.cs

```diff
@@ -52,6 +52,17 @@ public BodyModelBinderProvider(IList<IInputFormatter> formatters, IHttpRequestSt
             _loggerFactory = loggerFactory;
         }
 
+        /// <summary>
+        /// Gets or sets a flag that controls whether the binder should require non-null
+        /// input values to be supplied. Defaults to <c>true</c>.
+        /// </summary>
+        /// <example>
+        /// If <c>true</c>, and an associated <see cref="IInputFormatter"/> supplies <c>null</c>
+        /// as the value to bind (for example, if an incoming POST or PUT request supplies a
+        /// zero-length body), the binder will register a validation error.
+        /// </example>
+        public bool IsBindingRequired { get; set; } = true;
```

Comment:
Have changed it to a new option on `MvcOptions` called `AllowEmptyInputInInputFormatter`.

> Also add a property of this type to the InputFormatterContext class.

The new property on `InputFormatterContext` is just a `bool` flag called `AllowEmptyInput`, because `InputFormatterContext` is in `Abstractions` and hence doesn't know about `MvcOptions`.

> Instead have the input formatter(s) always return NoData in the empty body case. Copy GetDefaultValueForType() into BodyModelBinder and check the option there.

That would be a further change of design that ends up being a very subtle breaking change. In MVC 1.x, it's the input formatter subclasses that have the responsibility of selecting a default value for the model type if there's no input (by overriding `GetDefaultValueForType`). If we make them always return `NoData`, then they lose the ability to choose the default value. So what I've done instead is let the input formatters look at the `context.AllowEmptyInput` flag and the whole request context, and then each of them decides for itself whether to return `NoData` or `Success(defaultForModelType)`, since only the input formatter is able to supply a `defaultForModelType` (without the subtle breaking change).